crates/nu-command/src/network/url/join.rs

```diff
@@ -0,0 +1,342 @@
+                                left_span: value.expect_span(),
+                                right_message: format!("instead query is: {}", q),
+                                right_span: self
+                                    .query_span
+                                    .expect("query param span should be present"),
+                            });
+                        }
+                    }
+
+                    Ok(Self {
+                        query: Some(qs),
+                        params_span: Some(span),
+                        ..self
+                    })
+                }
+                Value::Error { error } => Err(error),
+                other => Err(ShellError::IncompatibleParametersSingle(
+                    String::from("Key params has to be a record"),
+                    other.expect_span(),
+                )),
+            };
+        }
+
+        // a part from port and params all other keys are strings.
+        match value.as_string() {
+            Ok(s) => {
+                if s.trim().is_empty() {
+                    Ok(self)
+                } else {
+                    match key.as_str() {
+                        "host" => Ok(Self {
+                            host: Some(s),
+                            ..self
+                        }),
+                        "scheme" => Ok(Self {
+                            scheme: Some(s),
+                            ..self
+                        }),
+                        "username" => Ok(Self {
+                            username: Some(s),
+                            ..self
+                        }),
+                        "password" => Ok(Self {
+                            password: Some(s),
+                            ..self
+                        }),
+                        "path" => Ok(Self {
+                            path: Some(if s.starts_with('/') {
+                                s
+                            } else {
+                                format!("/{}", s)
+                            }),
+                            ..self
+                        }),
+                        "query" => {
+                            if let Some(q) = self.query {
+                                if q != s {
+                                    // if query is present it means that also params_span is setted.
+                                    return Err(ShellError::IncompatibleParameters {
+                                        left_message: format!("Mismatch, query param is: {}", s),
+                                        left_span: value.expect_span(),
+                                        right_message: format!("instead qs from params is: {}", q),
+                                        right_span: self
+                                            .params_span
+                                            .expect("params span should be present"),
```

Comment:
I am not immediately convinced that this `.expect` will never panic based on the types going in (albeit sane on the first glance based on the logic). If you can prove this to be safe you could remove the wrapping `Option` or rely on either `Span::unknown` or our `Spanned<T>` wrapper type to link the span to its corresponding value in `UrlComponents`.